ssh/tailssh/incubator.go

```diff
@@ -268,53 +245,185 @@ func beIncubator(args []string) error {
-		if err != nil {
+	if attemptLoginShell {
+		// We weren't able to use login, maybe we can use su.
+		if handled, err := trySU(logf, ia); handled {
 			return err
+		} else {
+			logf("not attempting su")
 		}
-		groupIDs = append(groupIDs, int(gid))
 	}
 
-	if err := dropPrivileges(logf, ia.uid, ia.gid, groupIDs); err != nil {
+	// login and su didn't work, drop privileges and handle in-process.
+	if err := dropPrivileges(logf, ia); err != nil {
 		return err
 	}
 
 	if ia.isSFTP {
-		logf("handling sftp")
+		// Note - this does not trigger PAM authentication, so things like
+		// pam_mkhomedir won't work with FTP.
+		// TODO: it would be nice to make SFTP work with PAM, but that might
+		// require directly talking to the PAM API, which would require us to
+		// introduce CGO.
+		return handleSFTP(logf, ia)
+	}
 
-		server, err := sftp.NewServer(stdRWC{})
-		if err != nil {
-			return err
-		}
-		// TODO(https://github.com/pkg/sftp/pull/554): Revert the check for io.EOF,
-		// when sftp is patched to report clean termination.
-		if err := server.Serve(); err != nil && err != io.EOF {
-			return err
-		}
-		return nil
+	return handleSSHInProcess(logf, ia)
+}
+
+// handleSFTP serves SFTP connections.
+func handleSFTP(logf logger.Logf, ia incubatorArgs) error {
+	if err := dropPrivileges(logf, ia); err != nil {
+		return err
 	}
 
-	cmd := exec.Command(ia.cmdName, ia.cmdArgs...)
-	cmd.Stdin = os.Stdin
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	cmd.Env = os.Environ()
+	logf("handling sftp")
 
-	if ia.hasTTY {
-		// If we were launched with a tty then we should
-		// mark that as the ctty of the child. However,
-		// as the ctty is being passed from the parent
-		// we set the child to foreground instead which
-		// also passes the ctty.
-		// However, we can not do this if never had a tty to
-		// begin with.
-		cmd.SysProcAttr = &syscall.SysProcAttr{
-			Foreground: true,
+	server, err := sftp.NewServer(stdRWC{})
+	if err != nil {
+		return err
+	}
+	// TODO(https://github.com/pkg/sftp/pull/554): Revert the check for io.EOF,
+	// when sftp is patched to report clean termination.
+	if err := server.Serve(); err != nil && err != io.EOF {
+		return err
+	}
+	return nil
+}
+
+// shouldAttemptLoginShell decides whether we should attempt to get a full
+// login shell with the login or su commands.
+func shouldAttemptLoginShell(ia incubatorArgs) bool {
+	if ia.isSFTP {
+		return false
+	}
+
+	euid := os.Geteuid()
+	runningAsRoot := euid == 0
+
+	switch {
+	case !runningAsRoot:
+		// We have to be root in order to create a login shell.
+		return false
+	case hostinfo.IsSELinuxEnforcing():
+		// If we're running on a SELinux-enabled system, neiher login nor su
+		// will be able to set the correct context for the shell. So, we don't
+		// bother trying to run them and instead fall back to using the
+		// incubator to launch the shell.
+		// See http://github.com/tailscale/tailscale/issues/4908.
+		return false
+	}
+
+	return true
```

Comment:
nit: can this be `return runningAsRoot && !hostinfo.IsSELinuxEnforcing()`?

Reply:
Fixed.